tests/utils.py

```diff
@@ -154,69 +161,161 @@ class TestServerProcess():
+    Waits for the port message maximum timeout seconds.
+    """
+
+    self._start_process(extra_cmd_args, popen_cwd)
+    self._start_redirect_thread()
+
+    self._wait_for_port(timeout)
+
+    self.__logger.info(self.server + ' serving on ' + str(self.port))
+
+
+
+  def _start_process(self, extra_cmd_args, popen_cwd):
+    """Starts the process running the server."""
 
     # The "-u" option forces stdin, stdout and stderr to be unbuffered.
-    command = ['python', '-u', server, str(self.port)] + extra_cmd_args
+    command = ['python', '-u', self.server] + extra_cmd_args
 
-    # We are reusing one server subprocess in multiple unit tests, but we are
-    # collecting the logs per test.
+    # Reusing one subprocess in multiple tests, but split up the logs for each.
     self.__server_process = subprocess.Popen(command,
-        stdout=self.__temp_log_file, stderr=subprocess.STDOUT, cwd=popen_cwd)
+        stdout=subprocess.PIPE, stderr=subprocess.STDOUT, cwd=popen_cwd)
 
-    self.__logger.info('Server process with process id ' \
-        + str(self.__server_process.pid) + " serving on port " \
-        + str(self.port) + ' started.')
 
-    if timeout > 0:
-      try:
-        wait_for_server('localhost', self.server, self.port, timeout)
-      except Exception as e:
-        # Make sure that errors from the server side will be logged.
-        self.flush_log()
-        raise e
 
+  def _start_redirect_thread(self):
+    """Starts a thread responsible to redirect stdout/stderr to the Queue."""
 
+    # Run log_queue_worker() in a thread.
+    # The thread will exit when the child process dies.
+    self._log_queue = queue.Queue()
+    log_thread = threading.Thread(target=self._log_queue_worker,
+        args=(self.__server_process.stdout, self._log_queue))
 
-  def flush_log(self):
-    """Logs contents from TempFile, truncates buffer"""
+    # "daemon = True" means the thread won't interfere with the process exit.
+    log_thread.daemon = True
+    log_thread.start()
 
-    # Seek is needed to move the pointer to the beginning of the file, because
-    # the subprocess could have read and/or write and thus moved the pointer.
-    self.__temp_log_file.seek(0)
-    log_message = self.__temp_log_file.read()
 
-    if len(log_message) > 0:
-      title = "Test server (" + self.server + ") output:"
-      message = [title] + log_message.splitlines()
-      self.__logger.info('\n| '.join(message))
+  @staticmethod
+  def _log_queue_worker(stream, line_queue):
+    """
+    Worker function to run in a seprate thread.
+    Reads from 'stream', puts lines in a Queue (Queue is thread-safe).
+    """
 
-      # Make sure the file is empty before the next test logs new information.
-      self.__temp_log_file.truncate(0)
+    while True:
+      # readline() is a blocking operation.
+      # decode to push a string in the queue instead of 8-bit bytes.
+      log_line = stream.readline().decode()
```

Comment:
```suggestion
      log_line = stream.readline().decode('utf-8')
```
Wasn't the default value of 'utf-8' for encoding only added in Python 3? Throughout the rest of the code we consistently call `decode()` with an explicit encoding of 'utf-8' and that seems to be a fairly consistent pattern in my experience. Is that redundant?

Reply:
It seems that for python2 the default is `ascii`. 
It's better to specify the encoding while we have python2 code.
By the way: a good catch.